### ISSUE TITLE
Successfully running realtime inference on 6×H20, but quality issues remain: weak reasoning, slow barge-in, inaccurate ASR

### ISSUE THREAD

**Issue by candlewill** [MuyeHuang/DuplexOmni#5]
[duplexomni demo.mp3](https://github.com/user-attachments/files/30578439/duplexomni.demo.mp3)

I have successfully got DupleOmni realtime inference running on **6× H20 GPUs** with the following layout:

- GPU0: thinker  
- GPU1–4: talker main (TP=4)  
- GPU5: talker MTP

#### 1. Cap talker audio history (`server_talker.py`)

```text
TALKER_MAX_HISTORY_BLOCKS=20
```

- 1 block ≈ 480 ms of assistant audio  
- 20 blocks ≈ 9.6 s of audio context  
- Older blocks are trimmed so the talker prompt / prefill no longer grows without bound  

#### 2. Bound orchestrator queues (`server_orchestrator.py`)

```text
ORCH_MAX_TALKER_QUEUE_CHUNKS=2
ORCH_MAX_AUDIO_QUEUE_CHUNKS=4
```

- When the talker input queue is full, older chunks are dropped and only the latest ones are kept.  
- The same policy is applied to the audio output queue so stale audio is not played.  

Health endpoint now reports queue sizes and limits:

```json
{
  "talker_queue_size": 0,
  "audio_queue_size": 0,
  "max_talker_queue_chunks": 2,
  "max_audio_queue_chunks": 4
}
```

#### 3. Limit S1 history in the browser bridge (`omni_realtime_server.py`)

```text
MAX_S1_TURNS=20
```

#### 4. Keep only the latest user audio (`omni_realtime_server.py`)

```text
MAX_INPUT_QUEUE_CHUNKS=2
DROP_STALE_INPUT=1
```

If the microphone input queue backs up, older 480 ms chunks are discarded. Preferring the most recent audio improves perceived realtime behaviour.

### Remaining quality problems

The system is now able to run stably for longer sessions, but three functional issues are still prominent:

1. **Weak reasoning / off-topic answers**  
   The model frequently produces answers that do not address the user’s question (low “IQ”).

2. **Slow barge-in / interruption response**  
   When the user starts speaking while the assistant is talking, the system takes too long to stop and react.

3. **Inaccurate ASR**  
   Speech recognition errors are common and degrade the overall conversation quality.

### Request

Any guidance, known issues, or suggested directions for improving:

- reasoning quality / instruction following of the thinker,  
- barge-in latency,  
- ASR accuracy  

would be greatly appreciated.

**Comment by MuyeHuang** [comment]
Congratulations on getting the full pipeline running, and thanks for making all these improvements for long-session stability! Also, thank you very much for attaching such a detailed demo recording—it is genuinely helpful for locating the problems. After listening to the MP3, I noticed two fairly clear issues.

**1. Weak reasoning and off-topic responses**

DuplexOmni follows an S1–S2 design. S1 mainly focuses on fast and natural realtime interaction, while more complex questions are handled by S2 after S1 emits the corresponding trigger signal. Therefore, S1 alone is not expected to show particularly strong reasoning ability.

The off-topic responses are also clearly related to `MAX_S1_TURNS=20`. Each “turn” here corresponds to one 480 ms chunk, so 20 turns retain only about 9.6 seconds of context. This is particularly noticeable around the “说想法” (“talk about the idea”) part of the recording: the earlier context has already been truncated, so the model no longer knows which “idea” you are referring to. For a longer question, the beginning may already have been removed before the user finishes speaking.

**2. Slow barge-in**

After a clear user interruption, the model normally emits `[STOP]` within about 1–2 chunks, corresponding to roughly 480–960 ms at the model level. It would be useful to check exactly when `[STOP]` is generated. If it appears quickly but playback stops much later, the additional delay is more likely coming from inference backlog, networking, the server-side audio queue, or the client playback buffer. In addition, a very low user-input volume can reduce the model’s sensitivity to interruptions, causing it to detect the user’s barge-in later.

**3. ASR accuracy**

This is indeed a noticeable limitation of the current model. Its overall word error rate is roughly five times that of mainstream ASR models, and the error rate is even higher for short utterances. Purely streaming ASR is difficult because the model has to emit text immediately from each 480 ms chunk without much future context.

That said, in the uploaded MP3, the user-channel audio is also noticeably quiet and not very clear, which likely makes the ASR errors worse. We suggest increasing the loudness and clarity of the input audio before sending it to the model, while avoiding excessive amplification of background noise. Thanks again for providing the recording—it makes this issue much easier to diagnose.

One practical model-level direction is to introduce a pending-ASR mechanism: accumulate around 3–5 seconds of audio chunks and then emit the ASR result for the whole segment at once. This would require retraining, but it should provide a meaningful improvement in recognition accuracy.

One more thing: we have not trained or evaluated the model with such aggressive history truncation. Settings such as `MAX_S1_TURNS=20`, a very short Talker history, and dropping intermediate chunks may introduce additional degradation. To first measure the model itself, we suggest temporarily removing these short limits, disabling chunk dropping, and running an offline simulation that ignores wall-clock inference time. This should give a cleaner view of the model’s intrinsic streaming quality and its model-level response and barge-in latency measured in chunks.

We will continue improving and updating the model, but we do not have a firm timeline at the moment. Training a model like this is highly complex and requires substantial data preparation and computational resources.

**Comment by candlewill** [comment]
Thank you for the detailed analysis and for open-sourcing such a complete project — training data prep, training & inference code, and pretrained models. It’s the most complete realtime duplex system I’ve seen.

I’ll follow your suggestions: remove the aggressive history truncation / chunk dropping, improve input audio loudness, and re-run offline + realtime tests to better measure the model’s intrinsic quality and barge-in latency.

Regarding the pending-ASR idea, I’ll continue iterating on your codebase and also try finetuning on real conversational data.

Thanks again!

**Comment by chen821240134** [comment]
I also encountered the exact same problem as you did. The reasoning ability of the duplexomni model is too weak, and the voice announcement feels lacking in continuity.

**Comment by MuyeHuang** [comment]
> I also encountered the exact same problem as you did. The reasoning ability of the duplexomni model is too weak, and the voice announcement feels lacking in continuity.

Hi! Did you also use aggressive sequence or history truncation? Could you briefly share your configuration—especially the S1/Talker history limits, queue-dropping settings, and S2 model—so we can better identify the cause?
